src/v/storage/segment_utils.cc

```diff
@@ -402,26 +402,27 @@ ss::future<storage::index_state> do_copy_segment_data(
       seg->reader().filename(),
       tmpname);
     bool may_have_tombstone_records = false;
-    auto should_keep = [compacted_list = std::move(compacted_offsets),
-                        past_tombstone_delete_horizon,
-                        &may_have_tombstone_records](
-                         const model::record_batch& b,
-                         const model::record& r,
-                         bool) {
-        // Deal with tombstone record removal
-        if (r.is_tombstone() && past_tombstone_delete_horizon) {
-            return ss::make_ready_future<bool>(false);
-        }
+    auto should_keep =
+      [compacted_list = std::move(compacted_offsets),
+       past_tombstone_delete_horizon,
+       &may_have_tombstone_records,
+       &pb](const model::record_batch& b, const model::record& r, bool) {
+          // Deal with tombstone record removal
+          if (r.is_tombstone() && past_tombstone_delete_horizon) {
+              pb.add_removed_tombstone();
```

Comment:
nit: not sure how likely it is to be problematic, but something to consider is maybe we should only update this if the compaction succeeds

Reply:
Discussed offline, we will live with the overcounting for now in the edge cases of compaction failing past this point (partition shutdown/movement)